File: Backend/agent_solver.py
from typing import List, Dict, Set

class WordlyAgentSolver:
    """
    DFS-based solver for Wordly game
    """
    
    def __init__(self, word_list: List[str]):
        self.all_words = [w.upper() for w in word_list]
        self.possible_words = self.all_words.copy()
        self.guesses: List[Dict] = []
        
    def add_guess(self, guess: str, evaluation: List[str]):
        """Add a guess and its evaluation to filter possible words"""
        self.guesses.append({
            'word': guess.upper(),
            'evaluation': evaluation
        })
        self._filter_words()
        
    def _filter_words(self):
        """Filter possible words based on all guesses"""
        filtered = []
        
        for word in self.possible_words:
            if self._matches_all_patterns(word):
                filtered.append(word)
        
        self.possible_words = filtered
        
    def _matches_all_patterns(self, word: str) -> bool:
        """Check if word matches all guess patterns"""
        for guess_data in self.guesses:
            if not self._matches_pattern(word, guess_data['word'], guess_data['evaluation']):
                return False
        return True
    
    def _matches_pattern(self, word: str, guess: str, evaluation: List[str]) -> bool:
        """Check if word matches a single guess pattern"""
        word_letters = list(word)
        guess_letters = list(guess)
        
        # Check correct positions (green)
        for i in range(5):
            if evaluation[i] == 'correct':
                if word_letters[i] != guess_letters[i]:
                    return False
        
        # Check present letters (yellow)
        for i in range(5):
            letter = guess_letters[i]
            
            if evaluation[i] == 'present':
                # Letter must be in word but not at this position
                if word_letters[i] == letter:
                    return False
                if letter not in word_letters:
                    return False
                    
            elif evaluation[i] == 'absent':
                # Letter should not be in word (unless correct/present elsewhere)
                has_correct_or_present = any(
                    guess_letters[j] == letter and evaluation[j] in ['correct', 'present']
                    for j in range(5)
                )
                if not has_correct_or_present and letter in word_letters:
                    return False
        
        return True
    
    def get_next_guess(self) -> Dict:
        """Get the next best guess using DFS strategy"""
        if len(self.possible_words) == 0:
            return {
                'guess': '',
                'confidence': 0.0,
                'remaining_words': 0
            }
        
        if len(self.possible_words) == 1:
            return {
                'guess': self.possible_words[0],
                'confidence': 1.0,
                'remaining_words': 1
            }
        
        # First guess: use optimal starter words
        if len(self.guesses) == 0:
            starters = ['AROSE', 'SLATE', 'CRATE', 'TRACE', 'STARE']
            for starter in starters:
                if starter in self.possible_words:
                    return {
                        'guess': starter,
                        'confidence': 0.8,
                        'remaining_words': len(self.possible_words)
                    }
        
        # Score and choose best word
        best_word = self._choose_best_word()
        confidence = 1.0 / len(self.possible_words) if len(self.possible_words) > 0 else 0.0
        
        return {
            'guess': best_word,
            'confidence': min(confidence * 5, 1.0),  # Scale confidence
            'remaining_words': len(self.possible_words)
        }
    
    def _choose_best_word(self) -> str:
        """Choose word that provides most information"""
        if len(self.possible_words) <= 1:
            return self.possible_words[0] if self.possible_words else ''
        
        # Calculate letter frequencies
        letter_freq: Dict[str, int] = {}
        position_freq: Dict[str, int] = {}
        
        for word in self.possible_words:
            for i, letter in enumerate(word):
                letter_freq[letter] = letter_freq.get(letter, 0) + 1
                key = f"{i}-{letter}"
                position_freq[key] = position_freq.get(key, 0) + 1
        
        # Score each word
        best_score = -1
        best_word = self.possible_words[0]
        
        # Only score first 100 words for performance
        sample_size = min(100, len(self.possible_words))
        
        for word in self.possible_words[:sample_size]:
            score = self._score_word(word, letter_freq, position_freq)
            if score > best_score:
                best_score = score
                best_word = word
        
        return best_word
    
    def _score_word(self, word: str, letter_freq: Dict[str, int], position_freq: Dict[str, int]) -> float:
        """Score a word based on information gain"""
        score = 0.0
        seen_letters: Set[str] = set()
        
        for i, letter in enumerate(word):
            # Letter frequency score
            if letter not in seen_letters:
                score += letter_freq.get(letter, 0)
                seen_letters.add(letter)
            
            # Position frequency score (weighted higher)
            key = f"{i}-{letter}"
            score += position_freq.get(key, 0) * 2
        
        return score
    
    def reset(self):
        """Reset solver state"""
        self.possible_words = self.all_words.copy()
        self.guesses = []

File: FrontEnd/src/utils/agent-solver.ts
import { LetterState } from "@/types/types";
import { evaluateGuess, normalize } from "./game-utils";

export interface GuessResult {
  word: string;
  evaluation: LetterState[];
}

export class WordlyAgentSolver {
  private wordList: string[];
  private currentGuesses: GuessResult[] = [];
  
  constructor(wordList: string[]) {
    this.wordList = wordList.map(w => normalize(w));
  }

  /**
   * DFS-based solver that finds the answer by systematically trying words
   * and filtering based on feedback
   */
  public async solveStep(answer: string): Promise<string | null> {
    // Filter possible words based on previous guesses
    let possibleWords = this.filterPossibleWords(this.wordList, this.currentGuesses);
    
    if (possibleWords.length === 0) {
      return null; // No valid words left
    }

    // Choose next word using DFS strategy
    const nextGuess = this.chooseBestGuess(possibleWords);
    
    // Simulate getting feedback
    const evaluation = evaluateGuess(answer, nextGuess);
    
    // Store this guess
    this.currentGuesses.push({
      word: nextGuess,
      evaluation
    });

    return nextGuess;
  }

  /**
   * Filter words that match all previous guess patterns
   */
  private filterPossibleWords(words: string[], guesses: GuessResult[]): string[] {
    return words.filter(word => {
      // Check if this word matches all previous guess patterns
      for (const guess of guesses) {
        if (!this.wordMatchesPattern(word, guess)) {
          return false;
        }
      }
      return true;
    });
  }

  /**
   * Check if a word matches the pattern from a previous guess
   */
  private wordMatchesPattern(word: string, guess: GuessResult): boolean {
    const { word: guessWord, evaluation } = guess;
    
    // Track letters we've seen
    const wordLetters = word.split('');
    const guessLetters = guessWord.split('');

    // First pass: Check "correct" positions
    for (let i = 0; i < guessLetters.length; i++) {
      if (evaluation[i] === 'correct') {
        if (wordLetters[i] !== guessLetters[i]) {
          return false;
        }
      }
    }

    // Second pass: Check "present" and "absent"
    for (let i = 0; i < guessLetters.length; i++) {
      const letter = guessLetters[i];
      
      if (evaluation[i] === 'present') {
        // Letter must be in word but not at this position
        if (wordLetters[i] === letter) {
          return false; // Same position
        }
        if (!wordLetters.includes(letter)) {
          return false; // Not in word at all
        }
      } else if (evaluation[i] === 'absent') {
        // Letter should not be in word (unless it's correct/present elsewhere)
        const letterHasCorrectOrPresent = evaluation.some((state, idx) => 
          guessLetters[idx] === letter && (state === 'correct' || state === 'present')
        );
        
        if (!letterHasCorrectOrPresent && wordLetters.includes(letter)) {
          return false;
        }
      }
    }

    return true;
  }

  /**
   * DFS strategy: Choose word that eliminates most possibilities
   * Uses frequency analysis and position information
   */
  private chooseBestGuess(possibleWords: string[]): string {
    if (possibleWords.length === 1) {
      return possibleWords[0];
    }

    // If this is the first guess, use a word with common letters
    if (this.currentGuesses.length === 0) {
      const commonStarters = ['AROSE', 'SLATE', 'CRATE', 'TRACE', 'STARE'];
      for (const starter of commonStarters) {
        if (possibleWords.includes(starter)) {
          return starter;
        }
      }
    }

    // Score each word based on letter frequency and position
    let bestWord = possibleWords[0];
    let bestScore = -1;

    for (const word of possibleWords.slice(0, Math.min(100, possibleWords.length))) {
      const score = this.scoreWord(word, possibleWords);
      if (score > bestScore) {
        bestScore = score;
        bestWord = word;
      }
    }

    return bestWord;
  }

  /**
   * Score a word based on how much information it provides
   */
  private scoreWord(word: string, possibleWords: string[]): number {
    let score = 0;
    const letters = word.split('');
    
    // Count letter frequencies in remaining possible words
    const letterFreq: Record<string, number> = {};
    const positionFreq: Record<string, number> = {};
    
    for (const possibleWord of possibleWords) {
      for (let i = 0; i < possibleWord.length; i++) {
        const letter = possibleWord[i];
        letterFreq[letter] = (letterFreq[letter] || 0) + 1;
        positionFreq[`${i}-${letter}`] = (positionFreq[`${i}-${letter}`] || 0) + 1;
      }
    }

    // Score based on frequency
    const uniqueLetters = new Set(letters);
    for (const letter of uniqueLetters) {
      score += letterFreq[letter] || 0;
    }

    // Bonus for position frequency
    for (let i = 0; i < letters.length; i++) {
      score += (positionFreq[`${i}-${letters[i]}`] || 0) * 2;
    }

    return score;
  }

  public reset(): void {
    this.currentGuesses = [];
  }

  public getGuessHistory(): GuessResult[] {
    return [...this.currentGuesses];
  }
}
